docs/architecture/adr-023-protobuf-naming.md

```diff
@@ -0,0 +1,238 @@
+* balance conciseness against the possibility of either over-optimizing (making
+names too short and cryptic) or under-optimizing (just accepting bloated names
+with lots of redundant information)
+
+These guidelines are meant to act as a style guide for both the SDK and
+third-party modules.
+
+As a starting point, we should adopt all of the [DEFAULT](https://buf.build/docs/lint-checkers#default)
+checkers in [Buf's](https://buf.build) including [`PACKAGE_DIRECTORY_MATCH`](https://buf.build/docs/lint-checkers#file_layout),
+except:
+* [PACKAGE_VERSION_SUFFIX](https://buf.build/docs/lint-checkers#package_version_suffix)
+* [SERVICE_SUFFIX](https://buf.build/docs/lint-checkers#service_suffix)
+
+Further guidelines to be described below.
+
+### Principles
+
+#### Concise and Descriptive Names
+
+Names should be descriptive enough to convey their meaning and distinguish
+them from other names.
+
+Given that we are using fully-qualifed names within
+`google.protobuf.Any` as well as within gRPC query routes, we should aim to
+keep names concise, without going overboard. The general rule of thumb should
+be if a shorter name would convey more or else the same thing, pick the shorter
+name.
+
+For instance, `cosmos.bank.Send` (16 bytes) conveys roughly the same information
+as `cosmos_sdk.x.bank.v1.MsgSend` (28 bytes) but is notably shorter.
+If we needed version 2 of bank, we could just write `cosmos.bank2.Send`.
+
+Such conciseness makes names both more pleasant to work with and take up less
+space within transactions and on the wire.
+
+We should also resist the temptation to over-optimize, by making names
+cryptically short with abbreviations. For instance, we shouldn't try to
+reduce `cosmos.bank.Send` to `csm.bk.Snd` just to save six bytes.
+
+The goal is to make names **_concise but not cryptic_**.
+
+#### Names are for Clients First
+
+Package and type names should be chosen for the benefit of users, not
+necessarily because of legacy concerns related to the go code-base.
+
+#### Plan for Longevity
+
+In the interests of long-term support, we should plan on the names we do
+choose to be in usage for a long time, so now is the opportunity to make
+the best choices for the future.
+
+### Versioning
+
+#### Don't Allow Breaking Changes in Stable Packages
+
+Always use a breaking change detector such as [Buf](https://buf.build) to prevent
+breaking changes in stable (non-alpha or beta) packages. Breaking changes can
+break smart contracts/persistent scripts and generally provide a bad UX for
+clients. With protobuf, there should usually be ways to extend existing
+functionality instead of just breaking it.
+
+#### Use Simple Version Suffixes
+
+Instead of using [Buf's recommended version suffix](https://buf.build/docs/lint-checkers#package_version_suffix),
+we can more concisely indicate version 2, 3, etc. of a package with a simple
+package suffix rather than a sub-package, ex. `cosmos.bank2.Send` or `cosmos.bank.Send2`.
```

Comment:
personal preference: I'm more in favor of explicitly version tagging, as opposed to suffixing the module name. Something like `cosmos.bank.v2.Send` and `cosmos.bank.Send` for v1.

If we do go for the suffix approach, I'd like to standardize no either suffixing the module, or message name. My preference is for `bank2.Send` as opposed to `bank.Send2`, as versioning will most often happen at the module level, as opposed to the individual message level, right?

Reply:
Well my logic there is that we don't need to create a sub-package just to indicate a version. But maybe it doesn't matter.

I'm not really sure if `bank2.Send` or `bank.Send2` will be more common as we haven't actually hit these cases yet. So it's all a bit hypothetical. Which is my mind is good reason to not use `.v1` as a starting point. In reality we can deal with v2 when we get there

Reply:
I have to agree with @clevinson here.

> Well my logic there is that we don't need to create a sub-package just to indicate a version. But maybe it doesn't matter.

In fact, this is exactly how Golang's modules does semantic versioning -- new (major) versions are included in new packages. 

I would opt for `cosmos.bank.v2.MsgSend`. Specifically, the version should be part of the path, where v1 is omitted just like Go modules.

Reply:
That's fine. I'll update.